### PR TITLE
Feature/aql subquery operations stack i research view executor

### PR DESCRIPTION
**Pull request by hkernbach**
New IResearchViewExecutor API

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8663/

**Comment by gnusi** [comment]
@hkernbach please assign me as the reviewer once it's done

**Comment by hkernbach** [comment]
Latest run: http://jenkins01:8080/job/arangodb-matrix-pr-linux/10458/

File: arangod/Aql/IResearchViewExecutor.cpp [hunk 1]
```diff
@@ -425,87 +428,107 @@ IResearchViewExecutorBase<Impl, Traits>::IResearchViewExecutorBase(
...
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_NOT_IMPLEMENTED);
+}
+
+template <typename Impl, typename Traits>
+std::tuple<ExecutorState, typename IResearchViewExecutorBase<Impl, Traits>::Stats, size_t, AqlCall>
+IResearchViewExecutorBase<Impl, Traits>::skipRowsRange(AqlItemBlockInputRange& inputRange,
+                                                       AqlCall& call) {
   TRI_ASSERT(_indexReadBuffer.empty());
   IResearchViewStats stats{};
+  bool offsetPhase = (call.getOffset() > 0);
```
Comment (Dronplane) [inline arangod/Aql/IResearchViewExecutor.cpp]:
```suggestion
  const bool offsetPhase = (call.getOffset() > 0);
```

File: arangod/Aql/IResearchViewExecutor.cpp [hunk 2]
```diff
@@ -1045,6 +1084,39 @@ size_t IResearchViewExecutor<ordered, materializeType>::skip(size_t limit) {
...
+    while (_itr->next()) {
+      skipped++;
+    }
+
+    ++_readerOffset;
+    _itr.reset();
+    _doc = nullptr;
+  }
+
+  saveCollection();
```
Comment (Dronplane) [inline arangod/Aql/IResearchViewExecutor.cpp]: If I got things correctly, after skipAll no produceRows possible, so saving collection there might be redundant

Comment (hkernbach) [reply arangod/Aql/IResearchViewExecutor.cpp]: Thanks. Done

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8711/

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 3]
```diff
@@ -134,17 +134,27 @@ constexpr bool is_one_of_v = (std::is_same_v<T, Es> || ...);
...
+    IResearchViewExecutor<false, arangodb::iresearch::MaterializeType::LateMaterialize>,
+    IResearchViewExecutor<false, arangodb::iresearch::MaterializeType::Materialize>,
+    IResearchViewExecutor<false, arangodb::iresearch::MaterializeType::NotMaterialize | arangodb::iresearch::MaterializeType::UseStoredValues>,
+    IResearchViewExecutor<false, arangodb::iresearch::MaterializeType::LateMaterialize | arangodb::iresearch::MaterializeType::UseStoredValues>,
+    IResearchViewExecutor<true, arangodb::iresearch::MaterializeType::NotMaterialize>,
+    IResearchViewExecutor<true, arangodb::iresearch::MaterializeType::LateMaterialize>,
+    IResearchViewExecutor<true, arangodb::iresearch::MaterializeType::Materialize>,
+    IResearchViewExecutor<true, arangodb::iresearch::MaterializeType::NotMaterialize | arangodb::iresearch::MaterializeType::UseStoredValues>,
+    IResearchViewExecutor<true, arangodb::iresearch::MaterializeType::LateMaterialize | arangodb::iresearch::MaterializeType::UseStoredValues>,
+    ShortestPathExecutor, EnumerateListExecutor, LimitExecutor>;
```
Comment (iurii-i-popov) [inline arangod/Aql/ExecutionBlockImpl.cpp]: Add IResearchViewMergeExecutor

Comment (hkernbach) [reply arangod/Aql/ExecutionBlockImpl.cpp]: Thanks for spotting. Added.

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/8715/

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 4]
```diff
@@ -134,17 +134,37 @@ constexpr bool is_one_of_v = (std::is_same_v<T, Es> || ...);
...
+    // CalculationExecutor<CalculationType::Condition>, CalculationExecutor<CalculationType::Reference>, CalculationExecutor<CalculationType::V8Condition>,
+    HashedCollectExecutor,
 #ifdef ARANGODB_USE_GOOGLE_TESTS
-                TestLambdaExecutor,
-                TestLambdaSkipExecutor,  // we need one after these to avoid compile errors in non-test mode
+    TestLambdaExecutor,
+    TestLambdaSkipExecutor,  // we need one after these to avoid compile errors in non-test mode
 #endif
-                TraversalExecutor, KShortestPathsExecutor, ShortestPathExecutor, EnumerateListExecutor, LimitExecutor>;
+    IResearchViewExecutor<false, arangodb::iresearch::MaterializeType::NotMaterialize>,
```
Comment (mchacki) [inline arangod/Aql/ExecutionBlockImpl.cpp]: lol... This template thing is getting out of hand ;)

File: arangod/Aql/IResearchViewExecutor.h [hunk 5]
```diff
@@ -285,30 +299,33 @@ class IResearchViewExecutorBase {
...
  private:
   bool next(ReadContext& ctx);
 
  protected:
   Infos const& _infos;
   Fetcher& _fetcher;
-  InputAqlItemRow _inputRow;
-  ExecutionState _upstreamState;
+  InputAqlItemRow _inputRow;      // Remove me after refactor is done
+  ExecutionState _upstreamState;  // Remove me after refactor is done
```
Comment (mchacki) [inline arangod/Aql/IResearchViewExecutor.h]: I think this one can really be removed.

File: arangod/Aql/IResearchViewExecutor.h [hunk 6]
```diff
@@ -285,30 +299,33 @@ class IResearchViewExecutorBase {
...
 
  private:
   bool next(ReadContext& ctx);
 
  protected:
   Infos const& _infos;
   Fetcher& _fetcher;
-  InputAqlItemRow _inputRow;
-  ExecutionState _upstreamState;
+  InputAqlItemRow _inputRow;      // Remove me after refactor is done
```
Comment (mchacki) [inline arangod/Aql/IResearchViewExecutor.h]: This one is still required, please remove the comment

**Comment by mchacki** [comment]
Jenkins again:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8756/

**Comment by mchacki** [comment]
Jenkins green

**Pull request merged**